resources/sdk/pureclouddotnet/templates/ApiClient.mustache

```diff
@@ -156,6 +169,44 @@ namespace {{packageName}}.Client
             return request;
         }
 
+        private bool HasAccessTokenExpired(int statusCode)
+        {
+            TimeSpan timeSinceEpoch = (DateTime.UtcNow - new DateTime(1970, 1, 1));
+            long timeElapsed = (long) (timeSinceEpoch.TotalSeconds) - TokenIssueTime;
+            return statusCode == 401 && timeElapsed > TokenTimeToLive;
+        }
+
+        private void HandleExpiredAccessToken()
+        {
+            if (Monitor.TryEnter(RefreshTokenLock, 0))
+            {
+                try
+                {
+                    AuthTokenInfo authTokenInfo = Extensions.AuthExtensions.PostToken(this, ClientId, ClientSecret, authorizationCode: RefreshToken, isRefreshRequest: true);
+                    Configuration.Default.AccessToken = authTokenInfo.AccessToken;
+                }
+                catch (Exception e)
+                {
+                    throw new ApiException(500, e.Message);
+                }
+                finally
+                {
+                    Monitor.Exit(RefreshTokenLock);
+                }
+            }
+            else
+            {
+                // Abort with error if we have waited 2 seconds and refresh still isn't complete
+                if (!Monitor.TryEnter(RefreshTokenLock, TimeSpan.FromSeconds(2))) {
```

Comment:
Make this duration configurable and default to 10 seconds. For applications using a slow internet connection or if auth service is a little backed up, this could take longer.